In [1]:
## Importing basic libraries for extracting data

import pandas as pd
import numpy as np

In [2]:
## Here i have my dataset

movie = pd.read_csv('tmdb_5000_movies.csv')
credit = pd.read_csv('tmdb_5000_credits.csv')

In [3]:
movie.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [4]:
credit.head(1)

movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [5]:
## Let we merge the credit on mvoies dataset so that we can make the prediction easy

movies = movie.merge(credit,on='title')
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

In [6]:
movies.shape

(4809, 23)

In [7]:
movies.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count', 'movie_id', 'cast', 'crew'],
      dtype='object')

## Let we are gonna write the columns which we can use here and remaining we can drop it

In [8]:
# Genres
# movie_id
# keyword
# Overview
# title
# cast 
# crew

In [9]:
# Now we are taking only these columns for our model

movies = movies[['movie_id','genres','keywords','overview','title','cast','crew']]
movies.head(3)

movie_id                                             genres  \
0     19995  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1       285  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2    206647  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [10]:
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

## Data Preprocessing

Now we convert                                                                                                              
[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]                                                                                                       
This format to this                                                                                                             
['Action','Adventure','Fantacy','Scifi'....]                                                                                 
So for that atfirst we convert it using a python module ast.litral_eval to convert string of list to list                    

In [11]:
import ast
def convert(obj):
    l = []
    for i in ast.literal_eval(obj):
        l.append(i['name'])
        return l

In [12]:
movies['genres']=movies['genres'].apply(convert)

In [13]:
movies.head(3)

movie_id       genres                                           keywords  \
0     19995     [Action]  [{"id": 1463, "name": "culture clash"}, {"id":...   
1       285  [Adventure]  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2    206647     [Action]  [{"id": 470, "name": "spy"}, {"id": 818, "name...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [14]:
movies['keywords'].value_counts()

[]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  412
[{"id": 10183, "name": "independent film"}]                                                                                                                                                                                                                                                                                                                                                                     

### Now we will do same things on keywords

In [15]:
movies['keywords'] = movies['keywords'].apply(convert)
movies.head()

movie_id       genres          keywords  \
0     19995     [Action]   [culture clash]   
1       285  [Adventure]           [ocean]   
2    206647     [Action]             [spy]   
3     49026     [Action]       [dc comics]   
4     49529     [Action]  [based on novel]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [17]:
import ast
def convert2(obj):
    l = []
    counter = 0
    for i in ast.literal_eval(obj):
        if counter !=3:
            l.append(i['name'])
            counter+=1
        else:
            break
        return l

In [18]:
movies['cast'] = movies['cast'].apply(convert2)
movies.head()

movie_id       genres          keywords  \
0     19995     [Action]   [culture clash]   
1       285  [Adventure]           [ocean]   
2    206647     [Action]             [spy]   
3     49026     [Action]       [dc comics]   
4     49529     [Action]  [based on novel]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                      title               cast  \
0                                    Avatar  [Sam Worthington]   
1  Pirates of the Caribbean: At World's End      [Johnny Depp]   
2                                   Spectre     [Daniel Craig]   
3                     The Dark Knight Rises   [Christian Bale]   
4                               John Carter    [Taylor Kitsch]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [19]:
## Now we fetch director name form crew column

def fetch_obj(obj):
    l = []
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            l.append(i['name'])
            break
    return l

In [20]:
movies['crew'] = movies['crew'].apply(fetch_obj)
movies.head()

movie_id       genres          keywords  \
0     19995     [Action]   [culture clash]   
1       285  [Adventure]           [ocean]   
2    206647     [Action]             [spy]   
3     49026     [Action]       [dc comics]   
4     49529     [Action]  [based on novel]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                      title               cast  \
0                                    Avatar  [Sam Worthington]   
1  Pirates of the Caribbean: At World's End      [Johnny Depp]   
2                                   Spectre     [Daniel Craig]   
3                     The Dark Knight Rises   [Christian Bale]   
4                               John Carter    [Taylor Kitsch]   

                  crew  
0      [James Cameron]  
1     [Gore Verbinski]  
2         [Sam Mendes]  
3  [Christopher Nolan]  
4     [Andrew Stanton]

In [21]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

### Now we just convert overview column into list so we can concatinate it

In [22]:
movies.dropna(inplace=True)

In [23]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())

In [24]:
movies.head()

movie_id       genres          keywords  \
0     19995     [Action]   [culture clash]   
1       285  [Adventure]           [ocean]   
2    206647     [Action]             [spy]   
3     49026     [Action]       [dc comics]   
4     49529     [Action]  [based on novel]   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                      title               cast  \
0                                    Avatar  [Sam Worthington]   
1  Pirates of the Caribbean: At World's End      [Johnny Depp]   
2                                   Spectre     [Daniel Craig]   
3                     The Dark Knight Rises   [Christian Bale]   
4                               John Carter    [Taylor Kitsch]   

                  crew  
0      [James Cameron]  
1     [Gore Verbinski]  
2         [Sam Mendes]  
3  [Christopher Nolan]  
4     [Andrew Stanton]

### Now we just remove the spaces between names because to perform better performance

In [25]:
movies['keywords'] = movies['keywords'].apply(lambda x:[i.replace(" ","")for i in x])
movies['genres'] = movies['genres'].apply(lambda x:[i.replace(" ","")for i in x])
movies['cast'] = movies['cast'].apply(lambda x:[i.replace(" ","")for i in x])
movies['crew'] = movies['crew'].apply(lambda x:[i.replace(" ","")for i in x])

In [26]:
movies.head()

movie_id       genres        keywords  \
0     19995     [Action]  [cultureclash]   
1       285  [Adventure]         [ocean]   
2    206647     [Action]           [spy]   
3     49026     [Action]      [dccomics]   
4     49529     [Action]  [basedonnovel]   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                      title              cast  \
0                                    Avatar  [SamWorthington]   
1  Pirates of the Caribbean: At World's End      [JohnnyDepp]   
2                                   Spectre     [DanielCraig]   
3                     The Dark Knight Rises   [ChristianBale]   
4                               John Carter    [TaylorKitsch]   

                 crew  
0      [JamesCameron]  
1     [GoreVerbinski]  
2         [SamMendes]  
3  [ChristopherNolan]  
4     [AndrewStanton]

## Now we concat 'keywords','genres','cast','crew','overview' in one column

In [27]:
movies['tags'] = movies['genres'] + movies['keywords'] + movies['overview'] + movies['cast'] + movies['crew']
movies.head()

movie_id       genres        keywords  \
0     19995     [Action]  [cultureclash]   
1       285  [Adventure]         [ocean]   
2    206647     [Action]           [spy]   
3     49026     [Action]      [dccomics]   
4     49529     [Action]  [basedonnovel]   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                      title              cast  \
0                                    Avatar  [SamWorthington]   
1  Pirates of the Caribbean: At World's End      [JohnnyDepp]   
2                                   Spectre     [DanielCraig]   
3                     The Dark Knight Rises   [ChristianBale]   
4                               John Carter    [TaylorKitsch]   

                 crew                                               tags  
0      [JamesCameron]  [Action, cultureclash, In, the, 22nd, century,...  
1     [GoreVerbinski]  [Adventure, ocean, Captain, Barbossa,, long, b...  
2         [SamMendes]  [Action, spy, A, cryptic, message, from, Bond’...  
3  [ChristopherNolan]  [Action, dccomics, Following, the, death, of, ...  
4     [AndrewStanton]  [Action, basedonnovel, John, Carter, is, a, wa...

In [28]:
new_df = movies[['movie_id','title','tags']]
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [Action, cultureclash, In, the, 22nd, century,...  
1  [Adventure, ocean, Captain, Barbossa,, long, b...  
2  [Action, spy, A, cryptic, message, from, Bond’...  
3  [Action, dccomics, Following, the, death, of, ...  
4  [Action, basedonnovel, John, Carter, is, a, wa...

### Converting list into strings

In [29]:
new_df['tags'] = new_df['tags'].apply(lambda x:" ".join(x))

C:\Users\ASUS\AppData\Local\Temp\ipykernel_11992\3089450492.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x:" ".join(x))


In [30]:
new_df['tags'][0]

'Action cultureclash In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. SamWorthington JamesCameron'

In [31]:
new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())
new_df.head()

C:\Users\ASUS\AppData\Local\Temp\ipykernel_11992\842480936.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())


movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  action cultureclash in the 22nd century, a par...  
1  adventure ocean captain barbossa, long believe...  
2  action spy a cryptic message from bond’s past ...  
3  action dccomics following the death of distric...  
4  action basedonnovel john carter is a war-weary...

## Now we should do TextVectorization on tags column
Machine learning algorithms most often take numeric feature vectors as input. Thus, when working with text documents, we need a way to convert each document into a numeric vector. This process is known as text vectorization. One common vectorization approach is to represent each document as a vector of word counts.                                                        
In this use Bag of Words (BoW) Term Frequency method

Before TextVectorization atfirst we will do:
## Now we perform Stemmer from nltk library
Stemmers pick up the inclusion of a core or stem word within a longer word. For example, a stemming algorithm might look at a word like “planning,” and correctly recognize that the root word or stem word is “plan.” This can be a helpful element of something that parses raw text for analysis

In [32]:
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [33]:
def stem(text):
    y=[]
    for i in text.split():
        y.append(ps.stem(i))
        
    return " ".join(y)

In [34]:
new_df['tags'] = new_df['tags'].apply(stem)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_11992\3213734980.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(stem)


In [35]:
new_df['tags'][0]

'action cultureclash in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. samworthington jamescameron'

### Now from here we use TextVectorization

In [36]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=5000, stop_words='english')

In [37]:
vectors = cv.fit_transform(new_df['tags']).toarray()
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [38]:
cv.get_feature_names()

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['000',
 '007',
 '10',
 '100',
 '10th',
 '11',
 '12',
 '12th',
 '13',
 '14',
 '15',
 '150',
 '16',
 '16th',
 '17',
 '17th',
 '18',
 '1890',
 '18th',
 '19',
 '1910',
 '1920',
 '1930',
 '1930s',
 '1937',
 '1941',
 '1944',
 '1945',
 '1950',
 '1950s',
 '1955',
 '1959',
 '1960',
 '1960s',
 '1962',
 '1964',
 '1965',
 '1967',
 '1970',
 '1970s',
 '1971',
 '1972',
 '1973',
 '1974',
 '1976',
 '1977',
 '1979',
 '1980',
 '1980s',
 '1985',
 '1990',
 '1999',
 '19th',
 '20',
 '200',
 '2000',
 '2001',
 '2003',
 '2009',
 '20th',
 '21st',
 '23',
 '24',
 '25',
 '30',
 '300',
 '35',
 '3d',
 '40',
 '47',
 '50',
 '500',
 '60',
 '60s',
 '70',
 '80',
 'aaron',
 'aaroneckhart',
 'abandon',
 'abduct',
 'abil',
 'abl',
 'aboard',
 'abov',
 'abroad',
 'absolut',
 'abus',
 'abuse',
 'academ',
 'academi',
 'academy',
 'accept',
 'access',
 'accid',
 'accident',
 'acclaim',
 'accompani',
 'accomplish',
 'account',
 'accus',
 'ace',
 'achiev',
 'acquaint',
 'acquir',
 'act',
 'action',
 'activ',
 'activist',
 'activi

In [39]:
from sklearn.metrics.pairwise import cosine_similarity

In [40]:
similarity = cosine_similarity(vectors)

In [41]:
sorted(list(enumerate(similarity[0])),reverse=True,key=lambda x:x[1])[1:6]

[(524, 0.2608745973749755),
 (3399, 0.24748737341529164),
 (2941, 0.22613350843332272),
 (2672, 0.2004459314343183),
 (2089, 0.19611613513818404)]

In [42]:
def recommend(movie):
    movie_index = new_df[new_df['title'] == movie].index[0]
    distances = similarity[movie_index]
    movies_list = sorted(list(enumerate(distances)),reverse=True,key=lambda x:x[1])[1:6]
    
    for i in movies_list:
        print(new_df.iloc[i[0]].title)

In [43]:
recommend('Batman Begins')

Night at the Museum
Inkheart
The Secret Life of Walter Mitty
Three Kingdoms: Resurrection of the Dragon
Master and Commander: The Far Side of the World


In [44]:
import pickle

In [45]:
# pickle.dump(new_df,open('movies.pkl','wb'))

In [46]:
new_df['title']

0                                         Avatar
1       Pirates of the Caribbean: At World's End
2                                        Spectre
3                          The Dark Knight Rises
4                                    John Carter
                          ...                   
4801                                        Bang
4802                                      Primer
4804                                 El Mariachi
4806                   Signed, Sealed, Delivered
4808                           My Date with Drew
Name: title, Length: 4384, dtype: object

In [47]:
# pickle.dump(similarity,open('similarity.pkl','wb'))